pilot/pkg/xds/delta.go

```diff
@@ -372,16 +371,6 @@ func (s *DiscoveryServer) shouldRespondDelta(con *Connection, request *discovery
 		return true
 	}
 
-	// If there is mismatch in the nonce, that is a case of expired/stale nonce.
-	// A nonce becomes stale following a newer nonce being sent to Envoy.
-	// TODO: due to concurrent unsubscribe, this probably doesn't make sense. Do we need any logic here?
-	if request.ResponseNonce != "" && request.ResponseNonce != previousInfo.NonceSent {
-		deltaLog.Debugf("ADS:%s: REQ %s Expired nonce received %s, sent %s", stype,
-			con.conID, request.ResponseNonce, previousInfo.NonceSent)
-		xdsExpiredNonce.With(typeTag.Value(v3.GetMetricType(request.TypeUrl))).Increment()
-		return false
```

Comment:
This is a meaningless metric since spontaneous request can be sent for delta

Reply:
```	
// When the DeltaDiscoveryRequest is a ACK or NACK message in response
	// to a previous DeltaDiscoveryResponse, the response_nonce must be the
	// nonce in the DeltaDiscoveryResponse.
	// Otherwise (unlike in DiscoveryRequest) response_nonce must be omitted.
```

`request.ResponseNonce != ""` already rules out the "spontaneous request" case?

Reply:
`request.ResponseNonce != ""` does not mean it is a spontaneous request. 
Take a look at 
https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#incremental-xds

`Spontaneous [DeltaDiscoveryRequests](https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/discovery/v3/discovery.proto#envoy-v3-api-msg-service-discovery-v3-deltadiscoveryrequest) from the client. This can be done to dynamically add or remove elements from the tracked [resource_names](https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/discovery/v3/discovery.proto#envoy-v3-api-field-service-discovery-v3-discoveryrequest-resource-names) set. In this case [response_nonce](https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/discovery/v3/discovery.proto#envoy-v3-api-field-service-discovery-v3-discoveryrequest-response-nonce) must be omitted.
`

Reply:
With more dataplanes, like https://github.com/kmesh-net/kmesh communicating with istiod, this could happen. So better both side respect xds protocol.

Reply:
I am not sure I understand. The old code was "if its not a spontaneous request and its stale, report metric".

My understanding is you are saying "This is not valid since it can apply to spontaneous request and is noisy".

From the spec:

> In this case [spontaneous request] response_nonce must be omitted

So I am not sure why you are saying this code is impacted by spontaneous requests

Reply:
> With more dataplanes, like https://github.com/kmesh-net/kmesh communicating with istiod, this could happen. So better both side respect xds protocol.

BTW: I am 100% on board with making sure we follow the spec properly so we support other clients like this. But I am not sure this is following the spec as I understand it

Reply:
>My understanding is you are saying "This is not valid since it can apply to spontaneous request and is noisy".

Yes, that can be nosiy if concurrent request with responseNonce

Reply:
For spontaneous concurrent requests and acks, it is very likely a nonce outdated.

Reply:
What do you mean panic? Its just a metric, and its supposed to capture that exact scenario

Reply:
Oh, just a metric and log, as it is debug will keep it .  panic is from the below snippet https://github.com/istio/istio/pull/48877/files#r1464180251.

Reply:
done